torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1751,16 +1773,30 @@ struct to_ir {
     };
     GatheredTypes gathered(typeParser_);
     gathered.gather(classinfo);
-    auto val = emitExpr(obj);
+    Value* val = emitExpr(obj);
     RefinementSet refinement;
     if (gathered.types.size() == 1 &&
         gathered.types.at(0)->isSubtypeOf(val->type()) &&
         obj.kind() == TK_VAR) {
       std::string ident = Var(obj).name().name();
-      Refinement isinstance(std::move(ident), gathered.types.at(0));
-      refinement = RefinementSet({isinstance}, {});
+      Refinement isinstance(ident, gathered.types.at(0));
+
+      if (val->type()->kind() == UnionType::Kind) {
+        auto union_type = val->type()->expect<UnionType>();
+        std::vector<TypePtr> not_isinstance_types =
```

Comment:
Isn't the code here when `isinstance` has worked successfully ? Don't we want all types in the Union that are subtypes of `isinstance`?

Reply:
The RefinementSet needs to have "true refinements" and "false refinements". I'm creating the "false refinements" here.